### ISSUE TITLE
Handling PeerDepedencies

### ISSUE THREAD

**Issue by gtrufitt** [guardian/source#334]
**Is your feature request related to a problem? Please describe.**
When installing src-components into consumers, peer dependencies easily fall out of sync because `0.x versioning doesn't match on major - it only exact matches. 

For example, DCR's `@guardian/src-button` dependency is currently is `0.17.0` whereas consent mangement platform is `0.16.1`. 

This does not match on the major because it's 0.x. 

![image](https://user-images.githubusercontent.com/638051/79772901-9af68880-8328-11ea-8ff5-2293e66b0546.png)

Major versions, however, match upwards on minors and patch.

E.g - React

`0.` versions

![image](https://user-images.githubusercontent.com/638051/79773269-22dc9280-8329-11ea-9ff5-8e599d58e081.png)

vs

`> 0.` versions

![image](https://user-images.githubusercontent.com/638051/79773386-515a6d80-8329-11ea-9647-9fc8058ae1e3.png)

The concern here is whether you believe the [Public API is considered stable](https://semver.org/#spec-item-4)? And is it something that can be handled with major bumps and `rc` releases as you're currently doing? 

I would say yes - I'm guessing `rc` releases are unusual on 0. alpha code as they signify significant public usage - and *any* release of 0. functionally acts the same as an `rc` release in the sense that it *exact matches* when installed.

**Describe the solution you'd like**
Bump src-foundations and components to `1.x`.

**Describe alternatives you've considered**
Desperately try to keep all the modules and consumers in sync.

**Comment by oliverlloyd** [comment]
👍 
src-foundation code is in production, being used by the browsers of real users so a `1.0` label seem fair

**Comment by SiAdcock** [comment]
Thanks for sharing this. I'm really concerned about this cascading problem with dependencies such as discussion-rendering and CMP bringing different `peerDependencies` up to a consumer application like dotcom-rendering. I understand your frustration and I want to do what I can to alleviate it.

I want to give you some context on how I've been working because I think the problem lies in the release process rather than the fact that Source is still on version `0.x`. I've been releasing new minor versions every two weeks. Each new release has included breaking API changes, and so necessitates a minor version upgrade.

## What's going on? 

The API is still undergoing some churn, mainly in Foundations (themes, typography, palette) but also a little bit in Button too. There have been changes to the way packages are exported, from CommonJS to ESM. And I anticipate more breaking changes in the next couple of releases too.

I've been trying to minimise the number of breaking changes per release to make upgrades manageable. This has also kept my maintenance overhead down, since I can simply keep merging pull requests into master and create a new release whenever I feel the time is right.

I think we're reaching the limits of sustainability with this approach.

## What can we do differently?

I think you'll understand that keeping the current release process and simply replacing minor version bumps with major ones won't fix the problem. It'll just shift the problem one number to the left as I make major bumps every two weeks, leading to version drift across dependent projects.

I'm also afraid to say that we will still continue making breaking API changes. Source is in production in several codebases, but we're still being adopted by new teams and environments. Inevitably, people will find new integration issues and we will have to respond to them.

We can change the release process, and make it less churny from a breaking change perspective. We could make small patch release more frequently...  peerDependencies don't have to match on patch versions, only minor versions. We could then save up a bunch of non-urgent breaking changes and release them all at once as a minor version bump once a month, or once every two months. 

This would mean less version bumping of consumable apps like CMP and discussion-rendering. It means slightly more maintenance overhead but I can find a way to make it work.

## But when are we going to go to version 1?

I find this a question of psychological safety more than anything. We still need to define what constitutes a stable (as opposed to experimental) component, something that is safe to start adopting. I think it would be premature to hit version 1 before we know what acceptable looks like. Once we have that definition, we can decide whether our components are meeting it, or close enough to meeting it, and make the major bump. 

This is on our backlog. It should probably be given a higher priority, but with such a small team it has been difficult to find that balance between improving process and governance and delivering features. I'll bring it up at our next planning meeting. 

Thanks for reading and for bearing with me. Please hit me back if you think I've missed anything, or if I've made any incorrect assumptions.

**Comment by gtrufitt** [comment]
Thanks for the thorough writeup here Simon! Really appreciate the explanation, it does make a lot of sense - I guess this is a consequence also of keeping the main and all sub-packages in sync, any breaking change in any package requires a breaking change in all.

> peerDependencies don't have to match on patch versions, only minor versions.

One thing to note, in a 0.x release - patch releases *aren't matched* on minor

![image](https://user-images.githubusercontent.com/638051/79838316-b4d8af80-83aa-11ea-9dd6-227f067fe5c0.png)

Unless I'm mistaken and the behaviour of peer dependencies is different to that calculator.

It's great to understand the thinking underneath, and I think things like Dependabot can help us manage this better, so maybe there's not an urgency beyond us all trying to keep on top of the dependencies while we're still in early stages of the design system.

**Comment by SiAdcock** [comment]
> patch releases aren't matched on minor

I think the carat is important:

![Screenshot 2020-04-21 at 09 53 04](https://user-images.githubusercontent.com/5931528/79846154-fe7ac780-83b5-11ea-8544-5819b01c661a.png)

The [semver range calculator](https://semver.npmjs.com/) is a great tool though, thanks for introducing me to it!

I'll keep this thread updated with changes to our process and a timescale for version 1 release. Thanks for raising this issue. It's not always easy to see the correct route, and asking these questions helps us to course correct when we're heading into potentially stormy waters 🙂

**Comment by gtrufitt** [comment]
> I think the carat is important

👍 👍 My bad!

**Comment by SiAdcock** [comment]
> I got confused by 0 releases not matching on 0 minor vs >1 minor!

Yes, that took me by surprise also! How can three little numbers be so confusing?

**Comment by sndrs** [comment]
have you thought about using something like https://github.com/semantic-release/semantic-release?

**Comment by sndrs** [comment]
semantic release [allows for sharable configs](https://semantic-release.gitbook.io/semantic-release/usage/shareable-configurations) – i wonder if using it (or something similar) for internally-directed packages would be useful? 

combining it with consumer repos using dependabot auto-merging + cypress + chromatic feels like you pretty much automate a merge in source master cascading out to CMP and DCR (for example) with high confidence and very little intervention?

**Comment by SiAdcock** [comment]
We're now at v1.0.0. I hope this helps a little with version management 🙂
